docs/source/reference/types.rst

```diff
@@ -0,0 +1,286 @@
+Types, Functions, and Interfaces
+================================
+
+The type system in Idris 2 is quite powerful, and is the main feature that sets it apart from simply typed languages like Haskell or OCaml.
+
+There are many ways to construct types, those being members of ``Type``, in Idris.
+
+-------------
+Functions
+-------------
+
+
+Function types, formed by ``(x : a) -> b``, are valid types, given that ``a`` and ``b`` are valid types
+As shorthand, for non-dependent types, ``a -> b`` is the same as ``(_ : a) -> b``.
+
+In addition, to do simple universal quantification, one can write ``forall x . a``, where ``x`` must be only a name (it can't have a type qualifier).
+This is desugared to ``{0 x : _} -> a``.
+
+^^^^^^^^^^^^^
+Implicits
+^^^^^^^^^^^^^
+
+Implicit function types, those where the value of a parameter can be inferred from the context, can take three forms.
+The first of these, ``{x : A} -> B``, forms the basis "basic" implicit types.
+That is, ``x`` should be inferred *syntactically*.
+This is the equivalent of implicit ``forall`` in Haskell or Rocq, where we might state that ``vectorMap`` on a vector might be:
+
+.. code:: idris
+
+    vectorMap : {n : Nat} -> (a -> b) -> Vect n a -> Vect n b
+
+We would then use ``vectorMap`` as follows:
+
+.. code-block:: idris
+
+    vectorMap f v
+
+or
+
+.. code:: idris
+
+    vectorMap {n = m} f v
+
+These two examples show the power of Idris implicits: while they can be inferred by the compiler, if necessary (or desired) they may be specified by the user
+
+Idris has *two* additional modifiers of implicits, however.
+The first, while not that important to the type system of Idris, is *incredibly* useful, and it is the default implicit.
+Essentially "regular" implicits only try to figure out the type without adding any extra information.
+
+Default implicits, on the other hand, instead of trying to figure out a value, will instead just use a default, which means that unnecessary "simple" functions are unnecessary in Idris: you just use the default value.
+
+The syntax for default implicits is ``{default val x : a} -> b``, where ``val : a``, which is the default value for ``x``.
+
+The third kind of implicits are auto implicits.
+Auto implicits are most useful in proofs, where they differ from regular implicits by there purpose.
```

Comment:
That's somewhat debatable. I suggest removing this sentence and instead show multiple uses for auto-implicits:

- proofs (like `Elem` or `IsJust`)
- interfaces
- effects (like how the compiler has tags to track allowable effects)